### ISSUE TITLE
Error: Stream ended prematurely when run jsipfs on node.js

### ISSUE THREAD

**Issue by kappabooom** [ipfs/js-ipfs#1033]
- **Version**:0.25.4
- **Platform**:macOSX
- **Subsystem**:

#### Type:
Question

#### Severity:
Medium
#### Description:
when I execute my code on node.js, I will get the error 

> Error: Stream ended prematurely
    at series (/Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/libp2p-secio/src/handshake/index.js:21:15)
    at /Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/internal/parallel.js:39:9
    at /Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/internal/once.js:12:16
    at iterateeCallback (/Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/internal/eachOfLimit.js:44:17)
    at /Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/internal/onlyOnce.js:12:16
    at /Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/internal/parallel.js:36:13
    at waterfall (/Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/libp2p-secio/src/handshake/propose.js:29:14)
    at /Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/internal/once.js:12:16
    at next (/Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/waterfall.js:21:29)
    at /Users/percomlab/Desktop/BIAU/GW2-ipfs/node_modules/async/internal/onlyOnce.js:12:16

>

**Comment by davidweisss** [comment]
I have the same error sporadically.

**Comment by diasdavid** [comment]
I've seen this as well, seems that SECIO is just a big leaker of uncaught errors. We get his one from the same place:

```
Error: MAC Invalid: b8dfdc3aec76c6ba59e0043a38303f2a4b5f118b0570b468054cda3987004c83 != 30956463f82f15a5439a941ec52fabfa82d8546c9647597cb15229562ed494ef
    at mac.digest (/Users/imp/code/js-ipfs/node_modules/libp2p-secio/src/etm.js:55:21)
    at Immediate.setImmediate (/Users/imp/code/js-ipfs/node_modules/libp2p-crypto/src/hmac/index.js:14:9)
    at runCallback (timers.js:785:20)
    at tryOnImmediate (timers.js:747:5)
    at processImmediate [as _immediateCallback] (timers.js:718:5)
```

//cc @dignifiedquire

**Comment by harrshasri** [comment]
Getting the Same Error

`Error: MAC Invalid: 1c6b78c202e07225ac16523f2f4ec15786763669924aaccb44e6c6f1ec0d10cd != 3667e92193877c2f25d36b7af2dad7d686722eddce9563b11725365ffee92b36
    at mac.digest (pricegraph.js:166183)
    at pricegraph.js:12370
    at run (pricegraph.js:87162)
    at runIfPresent (pricegraph.js:87191)
    at onGlobalMessage (pricegraph.js:87231)`

**Comment by diasdavid** [comment]
@harrshasri try updating to the latest version (current 0.28.2)

**Comment by harrshasri** [comment]
@diasdavid Ok Thanks

**Comment by diasdavid** [comment]
@jacobheun @alanshaw could you please confirm that this has been officially fixed?

**Comment by alanshaw** [comment]
As far as I can tell there's still and issue open for this - https://github.com/libp2p/js-libp2p-secio/issues/41

Does the fix need to come from libp2p-switch - should not be fatal error if a connection fails to handshake.

**Comment by jacobheun** [comment]
I can double check, but I believe the latest versions of libp2p-switch handle connection hang ups like these properly so they're not spitting out errors.

**Comment by jacobheun** [comment]
@diasdavid @alanshaw sorry for the delay, yes this can be closed. https://github.com/libp2p/js-libp2p-secio/pull/101 fixed this and libp2p/js-libp2p-secio#41 by ensuring errors are passed up the callback chain to the dialer.
